src/poly/evals.rs

```diff
@@ -52,11 +63,158 @@ where
+    ///
+    /// ## Mathematical Formula
+    ///
+    /// Given a polynomial $p(X_1, \ldots, X_n)$ represented by its evaluations, this
+    /// function computes the evaluations of the folded
+    /// polynomial $p'(X_2, \ldots, X_n) = p(r, X_2, \ldots, X_n)$.
+    ///
+    /// It uses the multilinear extension formula for the first variable:
+    ///
+    /// ```text
+    /// p(r, x') = p(0, x') + r \cdot (p(1, x') - p(0, x'))
+    /// ```
+    ///
+    /// where $x' = (x_2, \ldots, x_n)$ represents all other variables.
+    ///
+    /// ## Memory Access Pattern
+    ///
+    /// This function relies on the **lexicographical order** of the evaluation list.
+    /// - The first half of the slice contains all evaluations where $X_1 = 0$,
+    /// - The second half contains all evaluations where $X_1 = 1$.
+    ///
+    /// ```text
+    /// Before:
+    /// [ p(0, 0..0), p(0, 0..1), ..., p(0, 1..1) | p(1, 0..0), p(1, 0..1), ..., p(1, 1..1) ]
+    ///  └────────── Left Half (p(0, x')) ──────┘   └────────── Right Half (p(1, x')) ────┘
+    ///
+    /// After: (Computed in-place into the left half)
+    /// [ p(r, 0..0), p(r, 0..1), ..., p(r, 1..1) ]
+    ///   └───────── Folded result ─────────────┘
+    /// ```
+    ///
+    /// The function computes `result[i] = left[i] + r * (right[i] - left[i])` for
+    /// all `i` in the first half, and then truncates the list.
+    pub fn compress<F: Clone + Copy + Default + Send + Sync>(&mut self, r: F)
+    where
+        A: Algebra<F>,
+    {
+        assert_ne!(self.num_variables(), 0);
+        let num_evals = self.num_evals();
+        let mid = num_evals / 2;
+
+        // Evaluations at `a_i` and `a_{i + n/2}` slots are folded with `r` into `a_i` slot
+        let (p0, p1) = self.0.split_at_mut(mid);
+        if num_evals >= PARALLEL_THRESHOLD {
+            p0.par_iter_mut()
+                .zip(p1.par_iter())
+                .for_each(|(a0, &a1)| *a0 += (a1 - *a0) * r);
+        } else {
+            p0.iter_mut()
+                .zip(p1.iter())
+                .for_each(|(a0, &a1)| *a0 += (a1 - *a0) * r);
+        }
+        // Free higher part of the evaluations
+        self.0.truncate(mid);
     }
 
+    /// Folds a list of evaluations from a base field `F` into packed form of extension field `EF`.
+    ///
+    /// ## Arguments
+    /// * `r`: A value `r` from the extension field `EF`, used as the random challenge for folding.
+    ///
+    /// ## Returns
+    /// A new `EvaluationsList<EF>` containing the compressed evaluations in the extension field.
+    ///
+    /// The compression is achieved by applying the following formula to pairs of evaluations:
+    /// ```text
+    ///     p'(X_2, ..., X_n) = (p(1, X_2, ..., X_n) - p(0, X_2, ..., X_n)) \cdot r + p(0, X_2, ..., X_n)
+    /// ```
+    pub fn compress_packed<Ext>(&self, zi: Ext) -> EvaluationsList<Ext::ExtensionPacking>
+    where
+        A: Field,
+        Ext: ExtensionField<A>,
+    {
```

Comment:
To match the header documentation

```suggestion
    pub fn compress_packed<EF>(&self, r: EF) -> EvaluationsList<EF::ExtensionPacking>
    where
        A: Field,
        EF: ExtensionField<A>,
    {
```